lib/tools/adb-commands.js

```diff
@@ -14,4 +16,97 @@ methods.initZipAlign = async function() {
   this.binaries.zipalign = await this.getSdkBinaryPath("zipalign");
 };
 
+methods.getApiLevel = async function () {
+  log.info("Getting device API level");
+  try {
+    return this.shell(['getprop', 'ro.build.version.sdk']);
+  } catch (e) {
+    throw new Error(`Error getting device API level. Original error: ${e.message}`);
+  }
+};
+
+methods.isDeviceConnected = async function () {
+  let devices = await this.getConnectedDevices();
+  return devices.length > 0;
+};
+
+methods.mkdir = async function (remotePath) {
+  return this.shell(['mkdir', '-p', remotePath]);
+};
+
+methods.isValidClass = function (classString) {
+  // some.package/some.package.Activity
+  return new RegExp(/^[a-zA-Z0-9\./_]+$/).exec(classString);
+};
+
+methods.forceStop = async function (pkg) {
+  return this.shell(['am', 'force-stop', pkg]);
+};
+
+methods.clear = function (pkg) {
+  return this.shell(['pm', 'clear', pkg]);
+};
+
+methods.stopAndClear = async function (pkg) {
+  try {
+    await this.forceStop(pkg);
+    await this.clear(pkg);
+  } catch (e) {
+    throw new Error(`Cannot stop and clear ${pkg}. Original error: ${e.message}`);
+  }
+};
+
+methods.availableIMEs = async function () {
+  try {
+    let stdout = await this.shell(['ime', 'list', '-a']);
+    let engines = [];
+    for (let line of stdout.split('\n')) {
+      if (line.length > 0 && line[0] !== ' ') {
+        // remove newline and trailing colon, and add to the list
+        engines.push(line.trim().replace(/:$/, ''));
+      }
+    }
+    return engines;
+  } catch (e) {
+    throw new Error(`Error getting available IME's. Original error: ${e.message}`);
```

Comment:
the nice thing about logging them where we throw them is we get to see what happened just before. not just the overall failure without an idea of where in the log flow it actually took place